Improve networking checks

LukasReschke opened the pull request:
Whilst we currently state that [SSRF is generally outside of our threat model](https://docs.nextcloud.com/server/latest/admin_manual/installation/harden_server.html#ensure-that-your-nextcloud-instance-is-installed-in-a-dmz), this is something where we should invest to improve this.

Todos:

- [x] Make sure existing tests pass
- [x] Investigate how far tests can be added
- [x] Rebase and proper commit messages
- [x] phpcs

LukasReschke committed:
Improve networking checks

Whilst we currently state that SSRF is generally outside of our threat model, this is something where we should invest to improve this.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented on lib/private/Http/Client/Client.php:
```diff
@@ -57,17 +56,20 @@ class Client implements IClient {
 	private $logger;
 	/** @var ICertificateManager */
 	private $certificateManager;
+	private LocalAddressChecker $localAddressChecker;
```
Probably needs to be changed due to PHP version compatibility.

LukasReschke commented on lib/private/Http/Client/LocalAddressChecker.php:
```diff
@@ -0,0 +1,85 @@
+		}
+
+		// Disallow hostname only
+		if (substr_count($host, '.') === 0 && !(bool)filter_var($host, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
```
@nickvergessen FYI I changed this code here from 

```php
substr_count($host, '.') === 0
```

To this here. Otherwise it fails IPv6 HTTP URLs such as `http://[2001:4860:4860::8888]/`. - I hope that doesn't impact other things as well :-)

nickvergessen replied:
Right, seems the order of the blocks before was wrong

LukasReschke commented:
Couldn't test the app store page as it seems offline currently:

```
{"reqId":"MgQTR6OISoKtHPl9NqCw","level":2,"time":"2021-03-30T17:57:31+00:00","remoteAddr":"172.18.0.1","user":"--","app":"appstoreFetcher","method":"POST","url":"/index.php","message":"Could not connect to appstore: cURL error 7:  (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://apps.nextcloud.com/api/v1/apps.json","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15","version":"22.0.0.2"}
```

LukasReschke committed:
Adjust test

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke committed:
Generate autoload files

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

MorrisJobke commented on lib/private/Http/Client/DnsPinMiddleware.php:
```diff
@@ -0,0 +1,128 @@
+
+	public function __construct(
+		NegativeDnsCache $negativeDnsCache,
+		LocalAddressChecker $localAddressChecker) {
```
```suggestion
		LocalAddressChecker $localAddressChecker
	) {
```

MorrisJobke commented on lib/private/Http/Client/DnsPinMiddleware.php:
```diff
@@ -0,0 +1,128 @@
+					'443',
+				];
+
+				if ($port != null) {
```
If this is really only null then: 

```suggestion
				if ($port !== null) {
```

Because I think 0 or empty string is not a valid port, or?

MorrisJobke reviewed (approved):
Beside my nitpicks it looks good 👍

LukasReschke committed:
Strictify null check

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented:
Rebased and added the suggestions by Morris.

@rullzer @ChristophWurst - mind giving this a review as well? This will only land in master, so we can iron out any potential bugs before a release :)

kesselb commented on lib/private/Http/Client/LocalAddressChecker.php:
```diff
@@ -0,0 +1,85 @@
+		$this->logger = $logger;
+	}
+
+	public function ThrowIfLocalIp(string $ip) : void {
```
Should be `throwIfLocalIp`

MorrisJobke replied:
@LukasReschke 👀

sk1p commented:
Just a quick ping that this PR possibly (probably?) caused a regression for streaming from external storage, please see https://github.com/nextcloud/server/issues/30282
